### PR TITLE
Feature/109 CR fixes #2

### LINKED ISSUE #2: Use more suitable project template
Current cookiecutter template, https://github.com/ardydedase/cookiecutter-pypackage.git, lacks many of the things required to create a BibDB replacement. Instead of extending it with static files, proper application factory, CLI, endpoint tests et cetera, a dedicated Flask cookiecutter template should be put to use, e.g. [cookiecutter-flask](https://github.com/sloria/cookiecutter-flask).

### PR DESCRIPTION
**Pull request by mblomdahl**
Here is the other batch of fixes for #109 CR remarks.

It's incomplete, unfortunately. Because I got tired and my brain stopped working.. Practically speaking, all the missing features pointed out in #109 are fixed, except for the *"You will only see all permissions for those collections that you are cataloging administrator for."* notice. Unfortunately, since I'm lazy and always looking for shortcuts and "the easy way out", only some of the test coverage updates are in place. 🙀

@mxtthias If you have the time (and the pepp) tomorrow, it would be appreciated if you could add the suitable test codes.. (Marked with "raise NotImplementedError" exceptions.) If not, I'll get to it first thing  Monday morning. Should be done around lunch then.

### CODE REVIEW

File: tests/end2end/test_collection_view.py [hunk 1]
```diff
@@ -39,3 +43,100 @@ def test_user_sees_error_message_if_collection_code_does_not_exist(user, testapp
...
+    # Sees all permissions.
+    assert _('Cataloging Admins') in res
+    assert _('Permissions') in res
+    assert _('You will only see all permissions for those collections that you are '
+             'cataloging administrator for.')
+    assert other_users_non_cataloging_admin_permission.user.email not in res
+    assert other_user_with_cataloging_admin_permission.user.email in res
+
+
+def test_non_cataloging_admin_user_sees_permissions_table_on_collections_they_have_permissions_for(
```
Comment (mblomdahl) [inline tests/end2end/test_collection_view.py]: @mxtthias This needs to be fixed...

File: tests/end2end/test_collection_view.py [hunk 2]
```diff
@@ -39,3 +43,100 @@ def test_user_sees_error_message_if_collection_code_does_not_exist(user, testapp
...
+    # Submits.
+    res = form.submit().follow()
+    assert res.status_code is 200
+    # Goes to the right URL for viewing a collection.
+    res = testapp.get(url_for('collection.view', collection_code=other_collection.code))
+    assert res.status_code is 200
+    # Sees all permissions.
+    assert _('Cataloging Admins') in res
+    assert _('Permissions') in res
+    assert _('You will only see all permissions for those collections that you are '
```
Comment (mblomdahl) [inline tests/end2end/test_collection_view.py]: @mxtthias This fails.

File: tests/end2end/test_collection_view.py [hunk 3]
```diff
@@ -39,3 +43,100 @@ def test_user_sees_error_message_if_collection_code_does_not_exist(user, testapp
...
+    assert other_user_with_cataloging_admin_permission.user.email in res
+
+
+def test_non_cataloging_admin_user_sees_permissions_table_on_collections_they_have_permissions_for(
+        user, collection, testapp):
+    """View own and 'cataloging_admin' permissions on collection thyself is associated with."""
+    raise NotImplementedError('Fix me for PR #120!')
+
+
+def test_non_cataloging_admin_users_sees_only_cataloging_admins_on_unassociated_collections(
```
Comment (mblomdahl) [inline tests/end2end/test_collection_view.py]: @mxtthias This too needs to be fixed...

File: tests/models/test_collection.py [hunk 4]
```diff
@@ -103,6 +103,11 @@ def test_removing_permissions(superuser, user):
     assert permission not in collection.permissions
 
 
+def test_get_permissions_as_seen_by(user):
```
Comment (mblomdahl) [inline tests/models/test_collection.py]: @mxtthias Not much work, but my brain was fried..

File: tests/models/test_user.py [hunk 5]
```diff
@@ -171,6 +171,16 @@ def test_is_cataloging_admin(superuser, user):
     assert user.is_cataloging_admin is False
 
 
+def test_is_cataloging_admin_for(collection):
```
Comment (mblomdahl) [inline tests/models/test_user.py]: @mxtthias Same thing here.

File: tests/models/test_user.py [hunk 6]
```diff
@@ -171,6 +171,16 @@ def test_is_cataloging_admin(superuser, user):
...
+    """Test is_cataloging_admin_for return value."""
+    raise NotImplementedError('Fix me for PR #120!')
+
+
+def test_has_any_permission_for(collection):
```
Comment (mblomdahl) [inline tests/models/test_user.py]: @mxtthias And here.

File: tests/end2end/test_collection_view.py [hunk 7]
```diff
@@ -39,3 +43,100 @@ def test_user_sees_error_message_if_collection_code_does_not_exist(user, testapp
...
+
+
+def test_cataloging_admin_can_view_all_permissions_on_own_collection(user, collection, testapp):
+    """View all permissions on a collection that you're managing."""
+    # Add cataloging admin permission.
+    cataloging_admin_permission = PermissionFactory(user=user, collection=collection,
+                                                    cataloging_admin=True)
+    cataloging_admin_permission.save()
+    # Add 2nd cataloger+registrant permission.
+    other_users_non_cataloging_admin_permission = PermissionFactory(collection=collection,
```
Comment (mxtthias) [inline tests/end2end/test_collection_view.py]: Maybe we should add an extra permission with another collection here, to make it more clear that it's the collection that decides if we can see the permission or not? Or maybe that's covered sufficiently by other tests?

File: tests/end2end/test_user_view.py [hunk 8]
```diff
@@ -142,6 +142,6 @@ def test_user_can_only_view_cataloging_admin_permissions_on_others_when_not_admi
     res = testapp.get(
         url_for('user.view', user_id=another_users_non_cataloging_admin_permission.user.id))
     assert res.status_code is 200
-    assert _('You will only see permissions for those collections that you are cataloging '
+    assert _('You will only see all permissions for those collections that you are cataloging '
```
Comment (mxtthias) [inline tests/end2end/test_user_view.py]: Does this work? See above.

File: messages.pot [hunk 9]
```diff
@@ -301,7 +306,7 @@ msgid "View User '%(email)s'"
 msgstr ""
 
 #: tests/end2end/test_user_view.py:145 tests/models/test_user.py:234 xl_auth/user/models.py:186
-msgid "You will only see permissions for those collections that you are cataloging administrator for."
+msgid "You will only see all permissions for those collections that you are cataloging administrator for."
```
Comment (mxtthias) [inline messages.pot]: Maybe s/all/the/ ?

File: xl_auth/templates/collections/view.html [hunk 10]
```diff
@@ -32,62 +32,80 @@ <h1>{{ _('View Collection \'%(code)s\'', code=collection.code) }}</h1>
...
-                <th class="col-md-3 word-break-all">{{ _('Registrant') }}</th>
-                <th class="col-md-2 word-break-all">{{ _('Cataloger') }}</th>
-                <th class="col-md-3 word-break-all">{{ _('Cataloguing Administrator') }}</th>
-                <th class="col-md-2">{{ _('Created At') }}</th>
-            </tr>
-            </thead>
-            <tbody>
-            {% for permission in collection.permissions | sort(attribute='created_at') %}
+    {% if current_user.is_admin or current_user.is_cataloging_admin
+            or current_user.has_any_permission_for(collection) %}
```
Comment (mxtthias) [inline xl_auth/templates/collections/view.html]: I'm trying to figure this out. For a regular user looking at a collection, they will only see the Permissions table if they have at least one permission for that collection? And if that's the case, they will only see their own permissions (and those of the cataloging admins?)?

For the user view, I removed the bottom table in `simple_view.html` (which is shown to regular users). What do you think about doing the same for this? Regular users can see their own permissions by looking at their profile, so maybe we don't need to show them here too?

Comment (mblomdahl) [reply xl_auth/templates/collections/view.html]: > I'm trying to figure this out. For a regular user looking at a collection, they will only see the Permissions table if they have at least one permission for that collection? And if that's the case, they will only see their own permissions (and those of the cataloging admins?)?

Exactly.

> For the user view, I removed the bottom table in simple_view.html (which is shown to regular users). What do you think about doing the same for this? /.../

When viewing yourself, you see your own permissions. I think it makes sense to see the permissions for your "own" collections too. Or maybe it doesn't, but let's collect feedback from UX on this release and see what they think!

File: tests/end2end/test_user_view.py (hunk 8 above)
Comment (mblomdahl) [reply tests/end2end/test_user_view.py]: No, you're right.. Fixing it!

File: messages.pot (hunk 9 above)
Comment (mblomdahl) [reply messages.pot]: Was not thinking well. Fixing...

File: tests/end2end/test_collection_view.py (hunk 7 above)
Comment (mblomdahl) [reply tests/end2end/test_collection_view.py]: We're doing that in the new tests added in latest push I think, e.g. `test_cataloging_admin_sees_only_cataloging_admins_on_others_collection`. I think that it's pretty clear with the 6-7 tests on the same topic that I added today!

**Pull request merged**